File: 백준/Gold/16934. 게임 닉네임/게임 닉네임.cc
#include <iostream>

using namespace std;

const int ROOT = 1;
const int MX = 100000 * 10 + 2;
int unused = 2;
int ch[MX];
int nxt[MX][26];

int c2i(char c){
    return c - 'a';
}
void insert(string& s){
    int cur = ROOT;
    for(char c : s){
        if(nxt[cur][c2i(c)] == -1) 
            nxt[cur][c2i(c)] = unused++;
        cur = nxt[cur][c2i(c)];
    }
    ch[cur]++;
}

void printStarName(string& s){
    int cur = ROOT;
    string str = "";
    
    bool flag = true;
    for(char c : s){
        str += c;
        if(nxt[cur][c2i(c)] == -1){
            flag = false;
            break;
        } 
        cur = nxt[cur][c2i(c)];
    }    
    
    cout << str;
    if(flag && ch[cur]) cout << ch[cur] + 1;
    cout << "\n";
}

int main()
{

    ios_base::sync_with_stdio(false);
    cin.tie(0);

    int n;
    cin >> n;

    for(int i = 0; i < MX; i++){
        fill(nxt[i], nxt[i] + 26, -1);
    }
    
    for(int i = 0; i < n; i++){
        string v;
        cin >> v;
        printStarName(v);
        insert(v);
    }
    
    return 0;
}

File: 백준/Gold/16934. 게임 닉네임/게임 닉네임.java
import java.io.*;
import java.util.*;

public class Main {

    static int ROOT = 1;
    static int idx = 2;
    static final int MX = 100000 * 26 + 20;

    static int chk[] = new int[MX];

    static int nxt[][] = new int[MX][26];
    static String starName[] = new String[MX];

    static void insert(String str, int wordIdx) {
        int cur = ROOT;

        for (int i = 0; i < str.length(); i++) {
            char c = str.charAt(i);
            if (nxt[cur][c2i(c)] == 0) {
                if (starName[wordIdx] == null) {
                    starName[wordIdx] = str.substring(0, i + 1);
                }
                nxt[cur][c2i(c)] = idx++;
            }
            cur = nxt[cur][c2i(c)];
        }

        chk[cur] += 1;
        if (starName[wordIdx] == null) {
            if (chk[cur] >= 2) {
                str += String.valueOf(chk[cur]);
            }
            starName[wordIdx] = str;
        }
    }

    static int c2i(char c) {
        return c - 'a';
    }


    public static void main(String[] args) throws IOException {
      
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());

        for (int i = 0; i < N; i++) {
            insert(br.readLine(), i);
        }
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < N; i++) {
            sb.append(starName[i]).append("\n");
        }
        System.out.println(sb);
    }


}